sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3968,3 +3964,267 @@ object ArrayUnion {
+            }
+            (signature, CodeGenerator.boxedType(elementType),
+              CodeGenerator.javaType(elementType), value,
+              s"get$ptName($i)", s"set$ptName($pos, $value)", CodeGenerator.javaType(elementType),
+              s"""
+                 |${ctx.createUnsafeArray(unsafeArray, size, elementType, s" $prettyName failed.")}
+                 |${ev.value} = $unsafeArray;
+               """.stripMargin)
+          case _ =>
+            val genericArrayData = classOf[GenericArrayData].getName
+            val et = ctx.addReferenceObj("elementType", elementType)
+            ("", "Object", "Object", value,
+              s"get($i, $et)", s"update($pos, $value)", "Object",
+              s"${ev.value} = new $genericArrayData(new Object[$size]);")
+        }
+      } else {
+        ("", "", "", "", "", "", "", "")
+      }
+
+    nullSafeCodeGen(ctx, ev, (array1, array2) => {
+      if (openHashElementType != "") {
+        // Here, we ensure elementTypeSupportEquals is true
+        val notFoundNullElement = ctx.freshName("notFoundNullElement")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$openHashElementType()"
+        val hs = ctx.freshName("hs")
+        val arrayData = classOf[ArrayData].getName
+        val arrays = ctx.freshName("arrays")
+        val array = ctx.freshName("array")
+        val arrayDataIdx = ctx.freshName("arrayDataIdx")
+
+        val array2NullCheck = if (right.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+            |if ($array2.isNullAt($i)) {
+            |  $notFoundNullElement = false;
+            |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+        val array1NullCheck = if (left.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+             |if ($array1.isNullAt($i)) {
+             |  if ($notFoundNullElement) {
+             |    $size++;
+             |    $notFoundNullElement = false;
+             |  }
+             |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+        val array1NullAssignment = if (left.dataType.asInstanceOf[ArrayType].containsNull) {
+          s"""
+             |if ($array1.isNullAt($i)) {
+             |  if ($notFoundNullElement) {
+             |    ${ev.value}.setNullAt($pos++);
+             |    $notFoundNullElement = false;
+             |  }
+             |} else
+           """.stripMargin
+        } else {
+          ""
+        }
+
+        s"""
+           |$openHashSet $hs = new $openHashSet$postFix($classTag);
+           |boolean $notFoundNullElement = true;
+           |int $size = 0;
+           |for (int $i = 0; $i < $array2.numElements(); $i++) {
+           |  $array2NullCheck
+           |  {
+           |    $javaTypeName $value = $array2.$getter;
+           |    $hsJavaTypeName $hsValue = $genHsValue;
+           |    $hs.add$postFix($hsValue);
```

Comment:
why wrap them inside `{...}`?

Reply:
ah i see. Maybe a better way is
```
def withArray2NullCheck(body: String) = if (right.dataType.asInstanceOf[ArrayType].containsNull) {
  s"""
    |if ($array2.isNullAt($i)) {
    |  $notFoundNullElement = false;
    |} else {
    |  $body
    |}""".stripMargin
} else {
  body
}
```

Reply:
This part will be part of `else clause` of `if`` statement depends on `$array2NullCheck` (i.e. if `(right.dataType.asInstanceOf[ArrayType].containsNull)` is true).